presto-main/src/main/java/io/prestosql/operator/scalar/ArrayContainsSubarray.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.operator.scalar;
+
+import io.airlift.slice.Slice;
+import io.prestosql.spi.block.Block;
+import io.prestosql.spi.function.Convention;
+import io.prestosql.spi.function.Description;
+import io.prestosql.spi.function.OperatorDependency;
+import io.prestosql.spi.function.ScalarFunction;
+import io.prestosql.spi.function.SqlNullable;
+import io.prestosql.spi.function.SqlType;
+import io.prestosql.spi.function.TypeParameter;
+import io.prestosql.spi.type.StandardTypes;
+import io.prestosql.spi.type.Type;
+
+import java.lang.invoke.MethodHandle;
+
+import static io.prestosql.spi.function.InvocationConvention.InvocationArgumentConvention.BLOCK_POSITION;
+import static io.prestosql.spi.function.InvocationConvention.InvocationArgumentConvention.NEVER_NULL;
+import static io.prestosql.spi.function.InvocationConvention.InvocationReturnConvention.NULLABLE_RETURN;
+import static io.prestosql.spi.function.OperatorType.EQUAL;
+
+@Description("Determines whether given values exist in the array")
+@ScalarFunction("contains_subarray")
+public final class ArrayContainsSubarray
+{
+    private ArrayContainsSubarray() {}
+
+    @TypeParameter("T")
+    @SqlType(StandardTypes.BOOLEAN)
+    @SqlNullable
+    public static Boolean contains(
+            @TypeParameter("T") Type elementType,
+            @OperatorDependency(
+                    operator = EQUAL,
+                    argumentTypes = {"T", "T"},
+                    convention = @Convention(arguments = {BLOCK_POSITION, NEVER_NULL}, result = NULLABLE_RETURN))
```

Comment:
This code can be simplified if you ask for a method with `{BLOCK_POSITION, BLOCK_POSITION}` calling convention.  Then in the code below you can simply do:
```java
    @TypeParameter("T")
    @SqlType(StandardTypes.BOOLEAN)
    @SqlNullable
    public static Boolean containsAll(
            @TypeParameter("T") Type elementType,
            @OperatorDependency(
                    operator = IS_DISTINCT_FROM,
                    argumentTypes = {"T", "T"},
                    convention = @Convention(arguments = {BLOCK_POSITION, BLOCK_POSITION}, result = NULLABLE_RETURN))
                    MethodHandle distinctFrom,
            @SqlType("array(T)") Block arrayBlock,
            @SqlType("array(T)") Block value)
            throws Throwable
    {
        for (int valuePosition = 0; valuePosition < value.getPositionCount(); valuePosition++) {
            for (int arrayPosition = 0; arrayPosition < arrayBlock.getPositionCount(); arrayPosition++) {
                if ((boolean) distinctFrom.invokeExact(arrayBlock, arrayPosition, value, valuePosition)) {
                    return false;
                }
            }
        }
        return true;
    }
```

If you want, a contains subarray, then something like this would work (note there are better algorithms):
```java
@TypeParameter("T")
@SqlType(StandardTypes.BOOLEAN)
@SqlNullable
public static Boolean containsSubArray(
        @TypeParameter("T") Type elementType,
        @OperatorDependency(
                operator = IS_DISTINCT_FROM,
                argumentTypes = {"T", "T"},
                convention = @Convention(arguments = {BLOCK_POSITION, BLOCK_POSITION}, result = NULLABLE_RETURN))
                MethodHandle distinctFrom,
        @SqlType("array(T)") Block arrayBlock,
        @SqlType("array(T)") Block value)
        throws Throwable
{
    int arrayLimit = arrayBlock.getPositionCount() - value.getPositionCount();
    for (int arrayPosition = 0; arrayPosition < arrayLimit; arrayPosition++) {
        if (containsSubArrayAt(distinctFrom, arrayBlock, arrayPosition, value)) {
            return true;
        }
    }
    return false;
}

private static boolean containsSubArrayAt(MethodHandle distinctFrom, Block arrayBlock, int arrayPosition, Block value)
        throws Throwable
{
    for (int valuePosition = 0; valuePosition < value.getPositionCount(); valuePosition++) {
        if ((boolean) distinctFrom.invokeExact(arrayBlock, arrayPosition + valuePosition, value, valuePosition)) {
            return false;
        }
    }
    return true;
}
```

Reply:
Thanks for your guidance! I find that `java.util.Collections#indexOfSubList` could find subarray too,
```java
public static int indexOfSubList(List<?> source, List<?> target) {
        int sourceSize = source.size();
        int targetSize = target.size();
        int maxCandidate = sourceSize - targetSize;

        if (sourceSize < INDEXOFSUBLIST_THRESHOLD ||
            (source instanceof RandomAccess&&target instanceof RandomAccess)) {
        nextCand:
            for (int candidate = 0; candidate <= maxCandidate; candidate++) {
                for (int i=0, j=candidate; i<targetSize; i++, j++)
                    if (!eq(target.get(i), source.get(j)))
                        continue nextCand;  // Element mismatch, try next cand
                return candidate;  // All elements of candidate matched target
            }
        } else {  // Iterator version of above algorithm
           ...
        }
        return -1;  // No candidate matched the target
    }
```
Since JDK uses the same algorithm to achieve this, so I just pick up your code, but fix a little bug, `arrayPosition < arrayLimit` should be `arrayPosition <= arrayLimit`